The column name agr_id was not found in this ResultSet

# Bug Report

## Describe the Bug
Done the migration from the schema.sql file to my database.

### Expected Behavior
An error must not be there.

### Observed Behavior
`[2023-04-07 11:33:53] [SEVERE ] StateMachineManager [consumer-contract-negotiation] error caught
org.eclipse.edc.spi.persistence.EdcPersistenceException: org.postgresql.util.PSQLException: The column name agr_id was not found in this ResultSet.
	at org.eclipse.edc.sql.SqlQueryExecutor$1.tryAdvance(SqlQueryExecutor.java:123)
	at java.base/java.util.Spliterator.forEachRemaining(Spliterator.java:332)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at org.eclipse.edc.connector.store.sql.contractnegotiation.store.SqlContractNegotiationStore.lambda$nextForState$8(SqlContractNegotiationStore.java:178)
	at org.eclipse.edc.transaction.local.LocalTransactionContext.execute(LocalTransactionContext.java:74)
	at org.eclipse.edc.connector.store.sql.contractnegotiation.store.SqlContractNegotiationStore.nextForState(SqlContractNegotiationStore.java:172)
	at org.eclipse.edc.connector.contract.negotiation.ConsumerContractNegotiationManagerImpl.lambda$processNegotiationsInState$16(ConsumerContractNegotiationManagerImpl.java:359)
	at org.eclipse.edc.statemachine.StateProcessorImpl.process(StateProcessorImpl.java:46)
	at java.base/java.util.stream.ReferencePipeline$5$1.accept(ReferencePipeline.java:231)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.LongPipeline.reduce(LongPipeline.java:498)
	at java.base/java.util.stream.LongPipeline.sum(LongPipeline.java:456)
	at org.eclipse.edc.statemachine.StateMachineManager.performLogic(StateMachineManager.java:118)
	at org.eclipse.edc.statemachine.StateMachineManager.lambda$loop$2(StateMachineManager.java:106)
	at io.micrometer.core.instrument.composite.CompositeTimer.record(CompositeTimer.java:79)
	at io.micrometer.core.instrument.Timer.lambda$wrap$0(Timer.java:160)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: org.postgresql.util.PSQLException: The column name agr_id was not found in this ResultSet.
	at org.postgresql.jdbc.PgResultSet.findColumn(PgResultSet.java:3033)
	at org.postgresql.jdbc.PgResultSet.getString(PgResultSet.java:2888)
	at org.eclipse.edc.connector.store.sql.contractnegotiation.store.SqlContractNegotiationStore.mapContractAgreement(SqlContractNegotiationStore.java:290)
	at org.eclipse.edc.connector.store.sql.contractnegotiation.store.SqlContractNegotiationStore.extractContractAgreement(SqlContractNegotiationStore.java:333)
	at org.eclipse.edc.connector.store.sql.contractnegotiation.store.SqlContractNegotiationStore.mapContractNegotiation(SqlContractNegotiationStore.java:314)
	at org.eclipse.edc.sql.SqlQueryExecutor$1.tryAdvance(SqlQueryExecutor.java:120)
	... 29 more
`
## Steps to Reproduce
Steps to reproduce the behavior:
1. Add all tables from sql file to database
2. Execute the steps till contract negotiation
3. Now the error is there in logs.

--- a/extensions/control-plane/store/sql/contract-negotiation-store-sql/src/main/java/org/eclipse/edc/connector/store/sql/contractnegotiation/store/SqlContractNegotiationStore.java
+++ b/extensions/control-plane/store/sql/contract-negotiation-store-sql/src/main/java/org/eclipse/edc/connector/store/sql/contractnegotiation/store/SqlContractNegotiationStore.java
@@ -148,7 +148,7 @@ public class SqlContractNegotiationStore extends AbstractSqlStore implements Con
         return transactionContext.execute(() -> {
             try {
                 var statement = statements.createNegotiationsQuery(querySpec);
-                return executeQuery(getConnection(), true, this.contractNegotiationMapper(), statement.getQueryAsString(), statement.getParameters());
+                return executeQuery(getConnection(), true, contractNegotiationMapper(), statement.getQueryAsString(), statement.getParameters());
             } catch (SQLException e) {
                 throw new EdcPersistenceException(e);
             }
@@ -173,7 +173,7 @@ public class SqlContractNegotiationStore extends AbstractSqlStore implements Con
             var stmt = statements.getNextForStateTemplate();
             try (
                     var connection = getConnection();
-                    var stream = executeQuery(connection, true, this.contractNegotiationWithAgreementMapper(connection), stmt, state, clock.millis(), max)
+                    var stream = executeQuery(connection, true, contractNegotiationWithAgreementMapper(connection), stmt, state, clock.millis(), max)
             ) {
                 var negotiations = stream.collect(Collectors.toList());
                 negotiations.forEach(cn -> leaseContext.withConnection(connection).acquireLease(cn.getId()));
@@ -191,7 +191,7 @@ public class SqlContractNegotiationStore extends AbstractSqlStore implements Con
 
     private @Nullable ContractNegotiation findInternal(Connection connection, String id) {
         var sql = statements.getFindTemplate();
-        return executeQuerySingle(connection, false, this.contractNegotiationMapper(), sql, id);
+        return executeQuerySingle(connection, false, contractNegotiationMapper(), sql, id);
     }
 
     private void update(Connection connection, String negotiationId, ContractNegotiation updatedValues) {
@@ -310,7 +310,7 @@ public class SqlContractNegotiationStore extends AbstractSqlStore implements Con
     }
 
     private ResultSetMapper<ContractNegotiation> contractNegotiationMapper() {
-        return (resultSet -> mapContractNegotiation(resultSet, this::extractContractAgreement));
+        return resultSet -> mapContractNegotiation(resultSet, this::extractContractAgreement);
     }
 
     private ResultSetMapper<ContractNegotiation> contractNegotiationWithAgreementMapper(Connection connection) {

--- a/extensions/control-plane/store/sql/contract-negotiation-store-sql/src/main/java/org/eclipse/edc/connector/store/sql/contractnegotiation/store/SqlContractNegotiationStore.java
+++ b/extensions/control-plane/store/sql/contract-negotiation-store-sql/src/main/java/org/eclipse/edc/connector/store/sql/contractnegotiation/store/SqlContractNegotiationStore.java
@@ -24,6 +24,7 @@ import org.eclipse.edc.connector.store.sql.contractnegotiation.store.schema.Cont
 import org.eclipse.edc.spi.persistence.EdcPersistenceException;
 import org.eclipse.edc.spi.query.Criterion;
 import org.eclipse.edc.spi.query.QuerySpec;
+import org.eclipse.edc.sql.ResultSetMapper;
 import org.eclipse.edc.sql.lease.SqlLeaseContextBuilder;
 import org.eclipse.edc.sql.store.AbstractSqlStore;
 import org.eclipse.edc.transaction.datasource.spi.DataSourceRegistry;
@@ -86,9 +87,8 @@ public class SqlContractNegotiationStore extends AbstractSqlStore implements Con
     @Override
     public @Nullable ContractAgreement findContractAgreement(String contractId) {
         return transactionContext.execute(() -> {
-            var stmt = statements.getFindContractAgreementTemplate();
-            try {
-                return executeQuerySingle(getConnection(), true, this::mapContractAgreement, stmt, contractId);
+            try (var connection = getConnection()) {
+                return findContractAgreementInternal(connection, contractId);
             } catch (SQLException e) {
                 throw new EdcPersistenceException(e);
             }
@@ -148,7 +148,7 @@ public class SqlContractNegotiationStore extends AbstractSqlStore implements Con
         return transactionContext.execute(() -> {
             try {
                 var statement = statements.createNegotiationsQuery(querySpec);
-                return executeQuery(getConnection(), true, this::mapContractNegotiation, statement.getQueryAsString(), statement.getParameters());
+                return executeQuery(getConnection(), true, this.contractNegotiationMapper(), statement.getQueryAsString(), statement.getParameters());
             } catch (SQLException e) {
                 throw new EdcPersistenceException(e);
             }
@@ -173,7 +173,7 @@ public class SqlContractNegotiationStore extends AbstractSqlStore implements Con
             var stmt = statements.getNextForStateTemplate();
             try (
                     var connection = getConnection();
-                    var stream = executeQuery(connection, true, this::mapContractNegotiation, stmt, state, clock.millis(), max)
+                    var stream = executeQuery(connection, true, this.contractNegotiationWithAgreementMapper(connection), stmt, state, clock.millis(), max)
             ) {
                 var negotiations = stream.collect(Collectors.toList());
                 negotiations.forEach(cn -> leaseContext.withConnection(connection).acquireLease(cn.getId()));
@@ -184,9 +184,14 @@ public class SqlContractNegotiationStore extends AbstractSqlStore implements Con
         });
     }
 
+    private ContractAgreement findContractAgreementInternal(Connection connection, String contractId) {
+        var stmt = statements.getFindContractAgreementTemplate();
+        return executeQuerySingle(connection, false, this::mapContractAgreement, stmt, contractId);
+    }
+
     private @Nullable ContractNegotiation findInternal(Connection connection, String id) {
         var sql = statements.getFindTemplate();
-        return executeQuerySingle(connection, false, this::mapContractNegotiation, sql, id);
+        return executeQuerySingle(connection, false, this.contractNegotiationMapper(), sql, id);
     }
 
     private void update(Connection connection, String negotiationId, ContractNegotiation updatedValues) {
@@ -304,14 +309,29 @@ public class SqlContractNegotiationStore extends AbstractSqlStore implements Con
         return format("Expected to find %d items, but found %d", expectedSize, actualSize);
     }
 
-    private ContractNegotiation mapContractNegotiation(ResultSet resultSet) throws SQLException {
+    private ResultSetMapper<ContractNegotiation> contractNegotiationMapper() {
+        return (resultSet -> mapContractNegotiation(resultSet, this::extractContractAgreement));
+    }
+
+    private ResultSetMapper<ContractNegotiation> contractNegotiationWithAgreementMapper(Connection connection) {
+        return (resultSet -> mapContractNegotiation(resultSet, rs -> {
+            var agreementId = rs.getString(statements.getContractAgreementIdFkColumn());
+            if (agreementId == null) {
+                return null;
+            } else {
+                return findContractAgreementInternal(connection, agreementId);
+            }
+        }));
+    }
+
+    private ContractNegotiation mapContractNegotiation(ResultSet resultSet, ResultSetMapper<ContractAgreement> agreementMapper) throws Exception {
         return ContractNegotiation.Builder.newInstance()
                 .id(resultSet.getString(statements.getIdColumn()))
                 .counterPartyId(resultSet.getString(statements.getCounterPartyIdColumn()))
                 .counterPartyAddress(resultSet.getString(statements.getCounterPartyAddressColumn()))
                 .protocol(resultSet.getString(statements.getProtocolColumn()))
                 .correlationId(resultSet.getString(statements.getCorrelationIdColumn()))
-                .contractAgreement(extractContractAgreement(resultSet))
+                .contractAgreement(agreementMapper.mapResultSet(resultSet))
                 .state(resultSet.getInt(statements.getStateColumn()))
                 .stateCount(resultSet.getInt(statements.getStateCountColumn()))
                 .stateTimestamp(resultSet.getLong(statements.getStateTimestampColumn()))

--- a/spi/control-plane/contract-spi/src/testFixtures/java/org/eclipse/edc/connector/contract/spi/testfixtures/negotiation/store/ContractNegotiationStoreTestBase.java
+++ b/spi/control-plane/contract-spi/src/testFixtures/java/org/eclipse/edc/connector/contract/spi/testfixtures/negotiation/store/ContractNegotiationStoreTestBase.java
@@ -563,6 +563,23 @@ public abstract class ContractNegotiationStoreTestBase {
     }
 
     @Test
+    @DisplayName("Verify that nextForState returns the agreement")
+    void nextForState_withAgreement() {
+        var contractAgreement = createContract(ContractId.createContractId(UUID.randomUUID().toString()));
+        var negotiation = createNegotiationBuilder(UUID.randomUUID().toString())
+                .contractAgreement(contractAgreement)
+                .state(ContractNegotiationStates.AGREED.code())
+                .build();
+
+        getContractNegotiationStore().save(negotiation);
+
+        var batch = getContractNegotiationStore().nextForState(ContractNegotiationStates.AGREED.code(), 1);
+
+        Assertions.assertThat(batch).hasSize(1).usingRecursiveFieldByFieldElementComparator().containsExactly(negotiation);
+
+    }
+
+    @Test
     void queryAgreements_noQuerySpec() {
         IntStream.range(0, 10).forEach(i -> {
             var contractAgreement = createContract(ContractId.createContractId(UUID.randomUUID().toString()));

--- a/spi/control-plane/contract-spi/src/testFixtures/java/org/eclipse/edc/connector/contract/spi/testfixtures/negotiation/store/ContractNegotiationStoreTestBase.java
+++ b/spi/control-plane/contract-spi/src/testFixtures/java/org/eclipse/edc/connector/contract/spi/testfixtures/negotiation/store/ContractNegotiationStoreTestBase.java
@@ -575,7 +575,7 @@ public abstract class ContractNegotiationStoreTestBase {
 
         var batch = getContractNegotiationStore().nextForState(ContractNegotiationStates.AGREED.code(), 1);
 
-        Assertions.assertThat(batch).hasSize(1).usingRecursiveFieldByFieldElementComparator().containsExactly(negotiation);
+        assertThat(batch).hasSize(1).usingRecursiveFieldByFieldElementComparator().containsExactly(negotiation);
 
     }
 

